qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -418,151 +449,215 @@ def run(self, dag: DAGCircuit) -> DAGCircuit:
                 )
-            elif isinstance(decomposition, DAGCircuit):
-                dag.substitute_node_with_dag(node, decomposition)
-            elif isinstance(decomposition, Operation):
-                dag.substitute_node(node, decomposition)
 
-        return dag
+            # now we are free to synthesize
+            else:
+                # this returns the synthesized operation and the qubits it acts on -- note that this
+                # may be different than the original qubits, since we may use auxiliary qubits
+                synthesized, used_qubits = self._synthesize_operation(node.op, qubits, tracker)
+
+            # if the synthesis changed the operation (i.e. it is not None), store the result
+            # and mark the operation qubits as used
+            if synthesized is not None:
+                synthesized_nodes[node] = (synthesized, used_qubits)
+                tracker.used(qubits)  # assumes that auxiliary are returned in the same state
+
+            # if the synthesis did not change anything, just update the qubit tracker
+            # other cases can be added: swaps, controlled gates (e.g. if control is 0), ...
+            else:
+                if node.op.name in ["id", "delay", "barrier"]:
+                    pass  # tracker not updated, these are no-ops
+                elif node.op.name == "reset":
+                    tracker.reset(qubits)  # reset qubits to 0
+                else:
+                    tracker.used(qubits)  # any other op used the clean state up
+
+        # we did not change anything just return the input
+        if len(synthesized_nodes) == 0:
+            return dag
+
+        # Otherwise we will rebuild with the new operations. Note that we could also
+        # check if no operation changed in size and substitute in-place, but rebuilding is
+        # generally as fast or faster, unless very few operations are changed.
+        out = dag.copy_empty_like()
+        index_to_qubit = dict(enumerate(dag.qubits))
+
+        for node in dag.topological_op_nodes():
+            if node in synthesized_nodes:
+                op, qubits = synthesized_nodes[node]
+                qargs = tuple(index_to_qubit[index] for index in qubits)
+                if isinstance(op, Operation):
+                    out.apply_operation_back(op, qargs, cargs=[])
+                    continue
+
+                if isinstance(op, QuantumCircuit):
+                    op = circuit_to_dag(op, copy_operations=False)
+
+                if isinstance(op, DAGCircuit):
+                    qubit_map = {
+                        qubit: index_to_qubit[index] for index, qubit in zip(qubits, op.qubits)
+                    }
+                    for sub_node in op.op_nodes():
+                        out.apply_operation_back(
+                            sub_node.op, tuple(qubit_map[qarg] for qarg in sub_node.qargs)
+                        )
```

Comment:
Should we also worry about `cargs` here?

Reply:
Yes we probably should, e.g. when unrolling custom definitions that have clbits

Reply:
Edit: it didn't work, there was just no test for this 🙂